File: XZTheme/2/Objective-C/Supporting/NSObject+XZTheme.m
//
//  NSObject+XZTheme.m
//  XZKit
//
//  Created by mlibai on 2017/10/14.
//

#import "NSObject+XZTheme.h"
#import "XZTheme.h"
#import "Example-Swift.h"
@import ObjectiveC;

static const void * const _nameOfAppliedTheme     = &_nameOfAppliedTheme;
static const void * const _needsThemeAppearanceUpdate  = &_needsThemeAppearanceUpdate;


@implementation NSObject (XZTheme)

- (XZTheme)xz_appliedTheme {
    return objc_getAssociatedObject(self, _nameOfAppliedTheme);
}

- (void)xz_setNeedsThemeAppearanceUpdate {
    if ([objc_getAssociatedObject(self, _needsThemeAppearanceUpdate) boolValue]) {
        return;
    }
    objc_setAssociatedObject(self, _needsThemeAppearanceUpdate, [NSNumber numberWithBool:YES], OBJC_ASSOCIATION_RETAIN);
    dispatch_async(dispatch_get_main_queue(), ^{
        [self xz_updateThemeAppearanceIfNeeded];
    });
}

- (void)xz_updateThemeAppearanceIfNeeded {
    if (![objc_getAssociatedObject(self, _needsThemeAppearanceUpdate) boolValue]) {
        return;
    }
    objc_setAssociatedObject(self, _needsThemeAppearanceUpdate, [NSNumber numberWithBool:NO], OBJC_ASSOCIATION_RETAIN);
    XZThemeStyle *themeStyle = [self.xz_themesIfLoaded styleForTheme:XZThemes.currentTheme];
    if (themeStyle != nil) {
        [self xz_updateAppearanceWithThemeStyle:themeStyle];
    }
    objc_setAssociatedObject(self, _nameOfAppliedTheme, XZThemes.currentTheme, OBJC_ASSOCIATION_COPY_NONATOMIC);
}

//- (void)xz_updateAppearanceWithThemeStyle:(XZThemeStyle *)themeStyle {
//    
//}

@end

File: XZTheme/2/Swift/Supporting/NSObject+XZTheme.swift
//
//  NSObject+XZTheme.swift
//  XZKit
//
//  Created by mlibai on 2017/12/10.
//  Copyright © 2017年 mlibai. All rights reserved.
//

import UIKit

extension NSObject {
    
    /// 更新主题外观：应用主题样式。当应用主题时，如果主题中有当前对象的主题样式，则此方法会被调用。
    ///
    /// - Parameter themeStyle: 待应用的主题样式。
    @objc(xz_updateAppearanceWithThemeStyle:) open func updateAppearance(with themeStyle: Theme.Style) {
       // 每个控件都有自己的最优实现方法。
    }
    
}
